src/lower/lowerer_impl_imperative.cpp

```diff
@@ -1668,17 +1670,42 @@ Stmt LowererImplImperative::lowerMergePoint(MergeLattice pointLattice,
       ir::Assign::make(indexSetIter.getCoordVar(), indexSetIter.getPosVar())
     );
     // Code to increment both iterator variables.
+    auto ivar = iter.getIteratorVar();
+    Expr iteratorParentPos = iter.getParent().getPosVar();
+    ModeFunction iterBounds = iter.posBounds(iteratorParentPos);
+    vector<Expr> iterGallopArgs = {
+      iter.getMode().getModePack().getArray(1),
+      ivar, iterBounds[1],
+      setMatch
+    };
+    auto indexVar = indexSetIter.getIteratorVar();
+    Expr indexIterParentPos = indexSetIter.getParent().getPosVar();
+    ModeFunction indexIterBounds = indexSetIter.posBounds(indexIterParentPos);
+    vector<Expr> indexGallopArgs = {
+      indexSetIter.getMode().getModePack().getArray(1),
+      indexVar, indexIterBounds[1],
+      setMatch
+    };
     auto incr = ir::Block::make(
-      compoundAssign(iter.getIteratorVar(), ir::Cast::make(Eq::make(iter.getCoordVar(), setMatch), iter.getIteratorVar().type())),
-      compoundAssign(indexSetIter.getIteratorVar(), ir::Cast::make(Eq::make(indexSetIter.getCoordVar(), setMatch), indexSetIter.getIteratorVar().type())),
+      ir::Assign::make(ivar, ir::Call::make("taco_gallop", iterGallopArgs, ivar.type())),
+      ir::Assign::make(indexVar, ir::Call::make("taco_gallop", indexGallopArgs, indexVar.type())),
       ir::Continue::make()
     );
     // Code that uses the defined parts together in the if-then-else.
     indexSetStmts.push_back(ir::IfThenElse::make(ir::And::make(iterEq, setEq), shiftDown, incr));
   }
 
   // Merge iterator coordinate variables
-  Stmt resolvedCoordinate = resolveCoordinate(mergers, coordinate, !resolvedCoordDeclared);
+  bool mergeWithMax;
```

Comment:
As mentioned above, this check can go earlier when the transformation is being applied, rather than at compilation time.